layout/FallbackQuery.js

```diff
@@ -0,0 +1,262 @@
+// This query is useful for specifying all the combinations of inputs starting
+// from most granular to least granular.  For example, given the input:
+// `30 w 26th street, new york, ny, usa`
+// the following pseudo-query would be generated:
+//
+// - (housenumber=30 && street=w 26th street && city=new york && state=ny && country=usa) ||
+// - (city=new york && state=ny && country=usa) ||
+// - (state=ny && country=usa) ||
+// - (country=usa)
+//
+// That is, it specifies as much as possible from the analyzed form, then falling
+// back in decreasing granularity.
+//
+// In the event of an input like `Socorro, Pennsylvania, Canada` where there is
+// no city named Socorro in Pennsylvania and there is no region named Pennsylvania
+// in Canada, the single result would be for country=Canada
+//
+// In the case that a full street+city+state+country is correct and found, all
+// OR'd queries will return results so only the most specific result should be
+// retained
+
+function Layout(){
+  this._score = [];
+}
+
+Layout.prototype.score = function( view, operator ){
+  this._score.push([ view, operator === 'must' ? 'must': 'should' ]);
+  return this;
+};
+
+function addPrimary(value, layer, fields, likely_to_have_abbreviation) {
+  // base primary query should match on layer and one of the admin fields via
+  // multi_match
+  var o = {
+    bool: {
+      must: [
+        {
+          term: { layer: layer }
```

Comment:
What should we do if the user also specified a `layers` parameter? It would be really easy for a lot of the queries to return no results if there are two conflicting layer inputs.